8240640: [macos] Wrong focus behaviour with multiple Alerts

pankaj-bansal opened the pull request:
The bug is a regression as a result of fix done for JDK-8227366 and is reproducible on Linux and Mac. This fix is being reverted in this change and a new bug (JDK-8271054) has been created to redo the JDK-8227366

An automated testcase is being added to make sure similar regression is not introduced when working on the redo bug. The automated testcase fails without the current change and passes after the change.

The testcase works fine on Mac and Linux, but I see some issues in windows. The stage is minimised after calling stage.show without calling stage.setAlwaysOnTop(true). I see that there are other tests like DualWindowsTest.java which are failing for same reason. Also, if I remove the stage.setAlwaysOnTop(true) from tests like FocusParentWindowOnChildCloseTest.java,  the stage remains minimised there as well. This seems wrong as calling stage.setAlwaysOnTop(true) should not be required to bring stage window in focus on running the test. This may be an issue on my Window machine or something similar. I hope reviewers can shed some light on this.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8240640](https://bugs.openjdk.java.net/browse/JDK-8240640): [macos] Wrong focus behaviour with multiple Alerts


### Reviewers
 * [Kevin Rushforth](https://openjdk.java.net/census#kcr) (@kevinrushforth - **Reviewer**)
 * [Ambarish Rapte](https://openjdk.java.net/census#arapte) (@arapte - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jfx pull/581/head:pull/581` \
`$ git checkout pull/581`

Update a local copy of the PR: \
`$ git checkout pull/581` \
`$ git pull https://git.openjdk.java.net/jfx pull/581/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 581`

View PR using the GUI difftool: \
`$ git pr show -t 581`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jfx/pull/581.diff">https://git.openjdk.java.net/jfx/pull/581.diff</a>

</details>

kevinrushforth commented on tests/system/src/test/java/test/robot/javafx/stage/WrongStageFocusWithApplicationModalityTest.java:
```diff
@@ -0,0 +1,114 @@
+            showAlert("Alert 3");
+        }
+
+        private void showAlert(String title)
```
Minor: curly brace on same line as method.

kevinrushforth commented on tests/system/src/test/java/test/robot/javafx/stage/WrongStageFocusWithApplicationModalityTest.java:
```diff
@@ -0,0 +1,114 @@
+            alert.show();
+        }
+    }
+}
```
Minor: add missing newline.

kevinrushforth commented:
/reviewers 2

kevinrushforth reviewed (commented):
The fix (reverting the earlier fix for JDK-8227366) looks fine to me. The test runs fine on Mac (and mostly fine on Linux, but I need to test a couple more things).

I also see the failure on Windows frequently depending on whether the Window initially gets focus. One thing you might try is to call `stage.toFront()` and `stage.requestFocus()` before showing the Alerts, although I'm not sure that will help. For other tests like this we do set the primary stage to be always on top, so that's probably the easiest. Failing that you could mark the test as unstable on Windows.

Btw, on my slow Linux VM, it occasionally fails because the window takes a bit too long to come up. I'll do some more testing once you fix the Windows issue.

I left a couple formatting comments on the test inline.

kevinrushforth commented on tests/system/src/test/java/test/robot/javafx/stage/WrongStageFocusWithApplicationModalityTest.java:
```diff
@@ -0,0 +1,114 @@
+
+    @Test(timeout = 25000)
+    public void testWindowFocusByClosingAlerts() throws Exception {
+        Thread.sleep(1000);
```
On my slower Linux machine if I change this to `sleep(3000)` is passes reliably. Given that you are opening up multiple dialogs, this seems like a reasonable change.

kevinrushforth commented on tests/system/src/test/java/test/robot/javafx/stage/WrongStageFocusWithApplicationModalityTest.java:
```diff
@@ -85,6 +89,16 @@ private static void keyPress(KeyCode code) throws Exception {
 
+    private static void mouseClick() {
+        Util.runAndWait(() -> {
+            robot.mouseMove((int) (alert.getX()+alert.getWidth()/2),
+                    (int) (alert.getY()+alert.getHeight()/2));
```
Minor: we usually put spaces around binary operators.

pankaj-bansal replied:
Fixed

kevinrushforth reviewed (approved):
The test changes look good (one minor formatting comment), and it now runs reliably for me on all three platforms.

pankaj-bansal commented:
> The fix (reverting the earlier fix for JDK-8227366) looks fine to me. The test runs fine on Mac (and mostly fine on Linux, but I need to test a couple more things).
> 
> I also see the failure on Windows frequently depending on whether the Window initially gets focus. One thing you might try is to call `stage.toFront()` and `stage.requestFocus()` before showing the Alerts, although I'm not sure that will help. For other tests like this we do set the primary stage to be always on top, so that's probably the easiest. Failing that you could mark the test as unstable on Windows.
> 
> Btw, on my slow Linux VM, it occasionally fails because the window takes a bit too long to come up. I'll do some more testing once you fix the Windows issue.
> 
> I left a couple formatting comments on the test inline.

I tried the stage.toFront or stage.requestFocus, they did not work. So I added the stage.setAlwaysOnTop(true). Along with that, I also added functionality to do mouse click at centre of alert to make sure keyboard focus is present on the alert. I have verified that the updated test fails before the fix and passes with the fix. I have tried multiple iterations of the test on all platforms (Mac 10.15, Ubuntu 20.04 and Windows 10) and test looks stable.

arapte reviewed (commented):
looks good to me.

pankaj-bansal commented:
/integrate
